### DISCUSSION TITLE
How to turn on logging on different threads properly?

### DISCUSSION THREAD

**Discussion by tisonkun** [tokio-rs/tracing#2618] [Q&A]
Sometimes, when I enable logging in the main function like:

```rust
tracing_subscriber::fmt()
        .with_max_level(LevelFilter::INFO)
        .init();
```

.. and spawn threads like:

```rust
let rpc = ServerRpc::new(peer, tx_requests);
std::thread::spawn(move || rpc.do_main());
```

Logging functions in `rpc.do_main` are not captured. But if I change it to:

```rust
{
            // start server rpc
            trace_span!("server rpc");
            let rpc = ServerRpc::new(peer, tx_requests);
            std::thread::spawn(move || rpc.do_main());
        }
```

Logging functions in `rpc.do_main` work properly.

Is there some best practices I understand this case? How can I properly config tracing manner and how can I add context kv like `slog` `Logger::new`.

FYI - I don't know what `trace_span` does, just randomly place some statements as a guess.

**Comment by davidbarsky** [comment]
> Logging functions in `rpc.do_main` are not captured. But if I change it to:

I'm not sure what you mean by this, but I'm guessing you're referring to the fact that logs in `rpc.do_main` don't have context.

> Is there some best practices I understand this case? How can I properly config tracing manner and how can I add context kv like `slog` `Logger::new`.
> FYI - I don't know what `trace_span` does, just randomly place some statements as a guess.

I would recommend reading [the documentation](https://docs.rs/tracing/latest/tracing/span/index.html) on spans which explains how they are used and why changing the code as you did makes things work. spans do not have a full 1:1 analogue to slog's `Logger`, but they are roughly similar.

**Reply by tisonkun** [reply]
> logs in rpc.do_main don't have context

No. It means the logs are not flush to the console, i.e., no output.
